src/libteam/patch/0013-add-support-for-custom-retry.patch

```diff
@@ -0,0 +1,358 @@
+ 				     * NULL in case this port is not selected */
+@@ -1110,6 +1153,7 @@ static int slow_addr_del(struct lacp_port *lacp_port)
+ #define LACP_SOCKET_CB_NAME "lacp_socket"
+ #define LACP_PERIODIC_CB_NAME "lacp_periodic"
+ #define LACP_TIMEOUT_CB_NAME "lacp_timeout"
++#define LACP_RETRY_COUNT_TIMEOUT_CB_NAME "lacp_retry_count_timeout"
+ 
+ static int lacp_port_timeout_set(struct lacp_port *lacp_port, bool fast_forced)
+ {
+@@ -1119,7 +1163,7 @@ static int lacp_port_timeout_set(struct lacp_port *lacp_port, bool fast_forced)
+ 
+ 	ms = fast_forced || lacp_port->lacp->cfg.fast_rate ?
+ 					LACP_PERIODIC_SHORT: LACP_PERIODIC_LONG;
+-	ms *= LACP_PERIODIC_MUL;
++	ms *= lacp_port->partner_retry_count;
+ 	ms_to_timespec(&ts, ms);
+ 	err = teamd_loop_callback_timer_set(lacp_port->ctx,
+ 					    LACP_TIMEOUT_CB_NAME,
+@@ -1286,6 +1330,9 @@ static int lacp_port_set_state(struct lacp_port *lacp_port,
+ 		err = lacp_port_partner_update(lacp_port);
+ 		if (err)
+ 			return err;
++		lacp_port->lacp->cfg.retry_count = LACP_CFG_DFLT_RETRY_COUNT;
++		teamd_loop_callback_disable(lacp_port->ctx,
++					    LACP_RETRY_COUNT_TIMEOUT_CB_NAME, lacp_port->lacp);
+ 		lacp_port_timeout_set(lacp_port, true);
+ 		teamd_loop_callback_enable(lacp_port->ctx,
+ 					   LACP_TIMEOUT_CB_NAME, lacp_port);
+@@ -1392,12 +1439,19 @@ static int lacpdu_send(struct lacp_port *lacp_port)
+ 	if (hwaddr_len != ETH_ALEN)
+ 		return 0;
+ 
+-	lacpdu_init(&lacpdu);
++	if (lacp_port->lacp->cfg.retry_count != LACP_CFG_DFLT_RETRY_COUNT || lacp_port->partner_retry_count != LACP_CFG_DFLT_RETRY_COUNT)
++		lacpdu_init(&lacpdu, 0xf1);
++	else
++		lacpdu_init(&lacpdu, 0x01);
+ 	lacpdu.actor = lacp_port->actor;
+ 	lacpdu.partner = lacp_port->partner;
+ 	memcpy(lacpdu.hdr.ether_shost, hwaddr, hwaddr_len);
+ 	memcpy(lacpdu.hdr.ether_dhost, ll_slow.sll_addr, ll_slow.sll_halen);
+ 	lacpdu.hdr.ether_type = htons(ETH_P_SLOW);
++	if (lacp_port->lacp->cfg.retry_count != LACP_CFG_DFLT_RETRY_COUNT || lacp_port->partner_retry_count != LACP_CFG_DFLT_RETRY_COUNT) {
++		lacpdu.v2.actor_retry_count = lacp_port->lacp->cfg.retry_count;
++		lacpdu.v2.partner_retry_count = lacp_port->partner_retry_count;
++	}
+ 
+ 	err = teamd_send(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0);
+ 	return err;
+@@ -1428,6 +1482,26 @@ static int lacpdu_process(struct lacp_port *lacp_port, struct lacpdu* lacpdu)
+ 			return err;
+ 	}
+ 
++	if (lacpdu->version_number == 0xf1) {
++		if (lacpdu->v2.actor_retry_count < 3) {
++			teamd_log_err("%s: retry count from partner (%u) out of its limits.", lacp_port->tdport->ifname, lacpdu->v2.actor_retry_count);
++			return -EINVAL;
++		}
++		if (lacp_port->partner_retry_count != lacpdu->v2.actor_retry_count)
++			teamd_log_dbg(lacp_port->ctx, "%s: retry count from partner changed from %u to %u",
++					   lacp_port->tdport->ifname,
++					   lacp_port->partner_retry_count,
++					   lacpdu->v2.actor_retry_count);
++		lacp_port->partner_retry_count = lacpdu->v2.actor_retry_count;
```

Comment:
Does this mean we allways expect the actor and partner to have same retry_count ?

Reply:
No. The actor fields in the `lacpdu` packet that we received refer to the partner fields in `lacp_port`. In other words, in the LACP PDU that we received, the actor information refers to our peer's information (aka our partner).